(window.webpackJsonp=window.webpackJsonp||[]).push([[108],{347:function(e,r,t){"use strict";t.r(r);var o=t(33),a=Object(o.a)({},(function(){var e=this,r=e.$createElement,t=e._self._c||r;return t("ContentSlotsDistributor",{attrs:{"slot-key":e.$parent.slotKey}},[t("h1",{attrs:{id:"nodejs-handling-errors"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#nodejs-handling-errors"}},[e._v("#")]),e._v(" nodejs - handling errors")]),e._v(" "),t("h2",{attrs:{id:"base-documentation"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#base-documentation"}},[e._v("#")]),e._v(" Base documentation")]),e._v(" "),t("p",[t("a",{attrs:{href:"https://nodejs.org/api/errors.html",target:"_blank",rel:"noopener noreferrer"}},[e._v("Node.js official documentation - errors - nodejs.org"),t("OutboundLink")],1)]),e._v(" "),t("p",[t("a",{attrs:{href:"https://www.joyent.com/node-js/production/design/errors",target:"_blank",rel:"noopener noreferrer"}},[e._v("Error Handling in Node.js - www.joyent.com"),t("OutboundLink")],1)]),e._v(" "),t("p",[t("a",{attrs:{href:"https://v8.dev/docs/stack-trace-api",target:"_blank",rel:"noopener noreferrer"}},[e._v("V8 stacktrace API - v8.dev"),t("OutboundLink")],1)]),e._v(" "),t("p",[t("a",{attrs:{href:"https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error",target:"_blank",rel:"noopener noreferrer"}},[t("code",[e._v("Error")]),e._v(" global object - developer.mozilla.org"),t("OutboundLink")],1)]),e._v(" "),t("p",[t("a",{attrs:{href:"https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch",target:"_blank",rel:"noopener noreferrer"}},[t("code",[e._v("try")]),e._v("..."),t("code",[e._v("catch")]),e._v(" - developer.mozilla.org"),t("OutboundLink")],1)]),e._v(" "),t("h2",{attrs:{id:"articles"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#articles"}},[e._v("#")]),e._v(" Articles")]),e._v(" "),t("p",[t("a",{attrs:{href:"https://goldbergyoni.com/checklist-best-practices-of-node-js-error-handling/",target:"_blank",rel:"noopener noreferrer"}},[e._v("Checklist: Best Practices of Node.JS Error Handling (2018)"),t("OutboundLink")],1)]),e._v(" "),t("h2",{attrs:{id:"tools"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#tools"}},[e._v("#")]),e._v(" Tools")]),e._v(" "),t("p",[t("a",{attrs:{href:"https://github.com/watson/stackman",target:"_blank",rel:"noopener noreferrer"}},[t("code",[e._v("watson/stackman")]),e._v(" - "),t("code",[e._v("github.com")]),t("OutboundLink")],1)]),e._v(" "),t("blockquote",[t("p",[e._v("Give Stackman an error and he will give an array of stack frames with extremely detailed information for each frame in the stack trace.")]),e._v(" "),t("p",[e._v("With Stackman you get access to the actual source code and surrounding lines for where the error occurred, you get to know if it happened inside a 3rd party module, in Node.js or in your own code.")]),e._v(" "),t("p",[e._v("By "),t("a",{attrs:{href:"https://github.com/watson",target:"_blank",rel:"noopener noreferrer"}},[e._v("Thomas Watson"),t("OutboundLink")],1),e._v(", dev at Elastic.")]),e._v(" "),t("p",[e._v("Project is backed by Elastic.")])]),e._v(" "),t("p",[t("a",{attrs:{href:"https://github.com/AriaMinaei/pretty-error",target:"_blank",rel:"noopener noreferrer"}},[t("code",[e._v("AriaMinaei/pretty-error")]),e._v(" - "),t("code",[e._v("github.com")]),t("OutboundLink")],1)]),e._v(" "),t("blockquote",[t("p",[e._v("See node.js errors with less clutter")])]),e._v(" "),t("p",[e._v("## Cores concepts")]),e._v(" "),t("h3",{attrs:{id:"sources"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#sources"}},[e._v("#")]),e._v(" sources")]),e._v(" "),t("p",[t("a",{attrs:{href:"https://www.joyent.com/node-js/production/design/errors",target:"_blank",rel:"noopener noreferrer"}},[e._v("Error Handling in Node.js - www.joyent.com"),t("OutboundLink")],1)]),e._v(" "),t("h3",{attrs:{id:"operational-errors-vs-programmer-errors"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#operational-errors-vs-programmer-errors"}},[e._v("#")]),e._v(" Operational errors vs programmer errors")]),e._v(" "),t("h4",{attrs:{id:"operational-errors"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#operational-errors"}},[e._v("#")]),e._v(" Operational errors")]),e._v(" "),t("p",[e._v("Represent run-time problems experienced by correctly-written programs. These are not bugs in the program.")]),e._v(" "),t("p",[e._v("Problem with :")]),e._v(" "),t("ul",[t("li",[e._v("the system itself (out of mem, too many open files, ...)")]),e._v(" "),t("li",[e._v("the system's configuration (no route to a remote host, ...)")]),e._v(" "),t("li",[e._v("the network (socket hangup, timeout, ...)")]),e._v(" "),t("li",[e._v("a remote service (500, failure to connect, ...)")])]),e._v(" "),t("h4",{attrs:{id:"programmer-errors"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#programmer-errors"}},[e._v("#")]),e._v(" Programmer errors")]),e._v(" "),t("p",[e._v("They are bugs. "),t("strong",[e._v("These are things that can always be avoided by changing the code.")])]),e._v(" "),t("p",[e._v("They are not handled correctly because the developer have not implemented error handling properly, not because it is impossible.")]),e._v(" "),t("p",[e._v("Ex :")]),e._v(" "),t("ul",[t("li",[e._v("tried to read property of "),t("code",[e._v("undefined")])]),e._v(" "),t("li",[e._v("async call without a callback")]),e._v(" "),t("li",[e._v("passed the wrong type in function arg (string instead of object, object instead of string, ...)")])]),e._v(" "),t("h4",{attrs:{id:"why-distinction-is-very-important"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#why-distinction-is-very-important"}},[e._v("#")]),e._v(" Why distinction is very important")]),e._v(" "),t("p",[e._v('People use the term "errors" to talk about both operational and programmer errors, but they\'re really quite different.')]),e._v(" "),t("p",[e._v("This distinction is very important: operational errors are part of the normal operation of a program. Programmer errors are bugs.")]),e._v(" "),t("p",[e._v("Sometimes, you have both operational and programming errors as part of the same root problem because they can chain themselves.")]),e._v(" "),t("p",[e._v("Example a server crashes because it try to use undefined variable, then clients will get an operational error (can't connect it is crashed).")]),e._v(" "),t("p",[e._v("Similarly, failure to handle an operational error is itself a programmer error.")]),e._v(" "),t("p",[e._v("The distinction between operational errors and programmer errors is the foundation for figuring out how to deliver errors and how to handle them.")]),e._v(" "),t("p",[e._v("### Handling errors")]),e._v(" "),t("h4",{attrs:{id:"handling-operational-errors"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#handling-operational-errors"}},[e._v("#")]),e._v(" Handling operational errors")]),e._v(" "),t("p",[e._v("Any code that does anything which might possibly fail (opening a file, connecting to a server, forking a child process, and so on) "),t("strong",[e._v("has")]),e._v(" to consider what happens when that operation fails.")]),e._v(" "),t("p",[e._v("That includes knowing "),t("strong",[e._v("how")]),e._v(" it may fail (the failure mode) and what such a failure would indicate.")]),e._v(" "),t("p",[e._v("The key point here is that error handling has to be done in a fine-grained way because the impact and response depend on exactly what failed and why")]),e._v(" "),t("p",[e._v("You may end up handling the same error at several levels of the stack. This happens when lower levels can't do anything useful except propagate the error to their caller, which propagates the error to its caller, and so on.")]),e._v(" "),t("p",[e._v("Often, only the top-level caller knows what the appropriate response is, whether that's to retry the operation, report an error to the user, or something else.")]),e._v(" "),t("p",[e._v("For any given error, there are a few things you might do:")]),e._v(" "),t("ul",[t("li",[e._v("Deal with the failure directly. (ex : database disconnect following a socket closed by database server)")]),e._v(" "),t("li",[e._v("Propagate the failure to your client. (ex : if user gave you invalid json, there is nothing you can do instead of telling him)")]),e._v(" "),t("li",[e._v("Retry the operation. (ex : error on a remote service, like 503 (unavailable), you may want to retry in a few seconds. This should be timebox (retry n times), documented and the end user should be notified)")]),e._v(" "),t("li",[e._v("Blow up. (the fail fast pattern, unexpected error, we fail as fast as we can because we can do nothing about it and the functionnal critical path of the app is down)")]),e._v(" "),t("li",[e._v("Log the error — and do nothing else. (ex : you can do nothing about the error, but there is no reason to crash because the error is not in the functional critical path of the app)")])]),e._v(" "),t("h4",{attrs:{id:"handling-programmer-errors"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#handling-programmer-errors"}},[e._v("#")]),e._v(" Handling programmer errors")]),e._v(" "),t("p",[t("strong",[e._v("The best way to recover from programmer errors is to crash immediately.")])]),e._v(" "),t("p",[e._v("You should run your programs using a restarter that will automatically restart the program in the event of a crash.")]),e._v(" "),t("p",[e._v("With a restarter in place, crashing is the fastest way to restore reliable service in the face of a transient programmer error.")]),e._v(" "),t("p",[e._v("The only downside to crashing on programmer errors is that connected clients may be temporarily disrupted, but remember :")]),e._v(" "),t("ul",[t("li",[e._v("By definition, these errors are "),t("strong",[e._v("always")]),e._v(" bugs.")]),e._v(" "),t("li",[e._v("The requests in flight are not necessarily going to complete successfully anyway.")]),e._v(" "),t("li",[e._v("In a reliable distributed system, clients must be able to deal with server failure by reconnecting and retrying requests.")]),e._v(" "),t("li",[e._v("If your production program is crashing so often that these disconnections are a problem, then the real problem is that the server is so buggy, not that it crashes in the case of a bug.")])]),e._v(" "),t("p",[e._v("The best way to debug these problems is to "),t("a",{attrs:{href:"https://www.joyent.com/node-js/production/debug#postmortem",target:"_blank",rel:"noopener noreferrer"}},[e._v("configure Node to dump core on an uncaught exception"),t("OutboundLink")],1),e._v(".")]),e._v(" "),t("p",[e._v("Finally, remember that a programmer error on a server just becomes an operational error on a client.")]),e._v(" "),t("h3",{attrs:{id:"patterns-for-writing-functions"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#patterns-for-writing-functions"}},[e._v("#")]),e._v(" Patterns for writing functions")]),e._v(" "),t("p",[e._v("If you don't know what errors can happen or don't know what they mean, then your program cannot be correct except by accident.")]),e._v(" "),t("p",[e._v("The question is : "),t("strong",[e._v("How do you deliver errors to the code that called your function?")])]),e._v(" "),t("h4",{attrs:{id:"throw-callback-reject-or-eventemitter"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#throw-callback-reject-or-eventemitter"}},[e._v("#")]),e._v(" Throw, Callback, Reject, or EventEmitter?")]),e._v(" "),t("h5",{attrs:{id:"the-basics"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#the-basics"}},[e._v("#")]),e._v(" The basics")]),e._v(" "),t("ul",[t("li",[t("code",[e._v("throw")]),e._v(" delivers an error synchronously. If the caller (or someone else in the call stack) used a "),t("code",[e._v("try")]),e._v(" / "),t("code",[e._v("catch")]),e._v(" then they catch the error. If none the program crashes (there is a particular case with domains and "),t("code",[e._v("uncaughtException")]),e._v(" event)")]),e._v(" "),t("li",[t("code",[e._v("callback(err, result)")]),e._v(", the historical pattern in node for asynchronous handling. The user passes the callback to your function. For him, "),t("code",[e._v("err")]),e._v(" or "),t("code",[e._v("result")]),e._v(" is "),t("code",[e._v("null")]),e._v(", never both "),t("code",[e._v("null")]),e._v(" or valued.")]),e._v(" "),t("li",[e._v("promise rejection, since node 8, upgrade of asynchronous handling to remove callback.")]),e._v(" "),t("li",[t("code",[e._v("async")]),e._v(" / "),t("code",[e._v("await")]),e._v(", since node 8, upgrade of asynchronous handling to look like synchronous. "),t("code",[e._v("try")]),e._v(" / "),t("code",[e._v("catch")]),e._v(" by the caller is supported.")]),e._v(" "),t("li",[t("code",[e._v("EventEmitter")]),e._v(" listening to "),t("code",[e._v("error")]),e._v(" event. To deal with streams. Function returns the "),t("code",[e._v("EventEmitter")]),e._v(" object. Caller listen to events.")])]),e._v(" "),t("h5",{attrs:{id:"when-do-you-use-throw-and-when-do-you-use-callbacks-or-event-emitters"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#when-do-you-use-throw-and-when-do-you-use-callbacks-or-event-emitters"}},[e._v("#")]),e._v(" When do you use "),t("code",[e._v("throw")]),e._v(", and when do you use callbacks or event emitters?")]),e._v(" "),t("p",[e._v("It depends :")]),e._v(" "),t("ul",[t("li",[e._v("Is the error an operational error or a programmer error?")]),e._v(" "),t("li",[e._v("Is the function itself synchronous or asynchronous?")])]),e._v(" "),t("h5",{attrs:{id:"the-most-common-case-is-an-operational-error-in-an-asynchronous-function"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#the-most-common-case-is-an-operational-error-in-an-asynchronous-function"}},[e._v("#")]),e._v(" The most common case is an operational error in an asynchronous function")]),e._v(" "),t("p",[e._v("By far, the most common case is an operational error in an asynchronous function (see "),t("code",[e._v("fs")]),e._v(" module for examples).")]),e._v(" "),t("h5",{attrs:{id:"the-next-most-common-case-is-an-operational-error-in-a-synchronous"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#the-next-most-common-case-is-an-operational-error-in-a-synchronous"}},[e._v("#")]),e._v(" The next most common case is an operational error in a synchronous")]),e._v(" "),t("p",[e._v("The next most common case is an operational error in a synchronous function like "),t("code",[e._v("JSON.parse")]),e._v(".\nFor these functions, if you encounter an operational error (like invalid user input), you have to deliver the error synchronously. You can either throw it (much more common) or return it.")]),e._v(" "),t("h5",{attrs:{id:"the-general-rule"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#the-general-rule"}},[e._v("#")]),e._v(" The general rule")]),e._v(" "),t("p",[e._v("For a given function, if any operational error can be delivered asynchronously, then "),t("strong",[e._v("all")]),e._v(" operational errors should be delivered asynchronously.")]),e._v(" "),t("p",[e._v("The general rule is that "),t("strong",[e._v("a function may deliver operational errors synchronously (e.g., by "),t("code",[e._v("throw")]),e._v("ing) or asynchronously (by passing them to a callback or emitting "),t("code",[e._v("error")]),e._v(" on an "),t("code",[e._v("EventEmitter")]),e._v("), but it should not do both.")])]),e._v(" "),t("p",[e._v("This way, a user can handle errors by either handling them in the callback or using "),t("code",[e._v("try")]),e._v("/"),t("code",[e._v("catch")]),e._v(", but "),t("strong",[e._v("they never need to do both")]),e._v(".")]),e._v(" "),t("h5",{attrs:{id:"programmer-errors-2"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#programmer-errors-2"}},[e._v("#")]),e._v(" Programmer errors")]),e._v(" "),t("p",[e._v("We've left out programmer errors. Recall that these are always bugs.")]),e._v(" "),t("p",[e._v("They can also usually be identified immediately by checking the types on arguments at the start of the function.")]),e._v(" "),t("p",[e._v("You should throw these errors immediately, since the program is broken. To do this, we recommend validating the types of all arguments at the start of the function.")]),e._v(" "),t("h4",{attrs:{id:"bad-input-programmer-error-or-operational-error"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#bad-input-programmer-error-or-operational-error"}},[e._v("#")]),e._v(" Bad input: programmer error or operational error?")]),e._v(" "),t("p",[e._v("How do you know what's a programmer error vs. an operational error? Quite simply: it's up to you to define and document what types your function will allow and how you'll try to interpret them.")]),e._v(" "),t("p",[t("strong",[e._v("If you get something other than what you've documented to accept, that's a programmer error.")])]),e._v(" "),t("p",[e._v("You have to use your judgment to decide how strict you want to be.")]),e._v(" "),t("h5",{attrs:{id:"example"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#example"}},[e._v("#")]),e._v(" Example")]),e._v(" "),t("p",[e._v("Imagine a function called "),t("code",[e._v("connect")]),e._v(" that takes an IP address and a callback and invokes the callback asynchronously after either succeeding or failing.")]),e._v(" "),t("p",[e._v("Suppose the user passes something that's obviously not a valid IP address, like "),t("code",[e._v("'bob'")]),e._v(".")]),e._v(" "),t("p",[e._v("In this case, you have a few options:")]),e._v(" "),t("ul",[t("li",[e._v("Document that the function only accepts strings representing valid IPv4 addresses, and throw an exception immediately if the user passes "),t("code",[e._v("'bob'")]),e._v(". This is strongly recommended.")]),e._v(" "),t("li",[e._v("Document that the function accepts any string. If the user passes "),t("code",[e._v("'bob'")]),e._v(", emit an asynchronous error indicating that you couldn't connect to IP address "),t("code",[e._v("'bob'")]),e._v(".")])]),e._v(" "),t("h5",{attrs:{id:"throwing-with-assert"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#throwing-with-assert"}},[e._v("#")]),e._v(" throwing with assert")]),e._v(" "),t("p",[e._v("Using "),t("a",{attrs:{href:"https://nodejs.org/api/assert.html",target:"_blank",rel:"noopener noreferrer"}},[t("code",[e._v("assert")]),e._v(" module"),t("OutboundLink")],1),e._v(" to check function parameters.")]),e._v(" "),t("p",[e._v("If the assertion is wrong, node throws.")]),e._v(" "),t("p",[e._v("Assert based exceptions "),t("a",{attrs:{href:"https://nodejs.org/api/errors.html#errors_exceptions_vs_errors",target:"_blank",rel:"noopener noreferrer"}},[e._v("are unrecoverable"),t("OutboundLink")],1),e._v(" at the JavaScript layer.")]),e._v(" "),t("p",[e._v("So it is an easy way to crash the app when it encounters a programmer error.")]),e._v(" "),t("h5",{attrs:{id:"strict-or-loose"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#strict-or-loose"}},[e._v("#")]),e._v(" Strict or loose ?")]),e._v(" "),t("p",[e._v("The better approach is to be super strict at the beginning and to be less strict in future versions without breaking what already exist.")]),e._v(" "),t("p",[e._v("If you do the opposite, super loose then going stricter, you will produce massives breaking changes to your users.")]),e._v(" "),t("h4",{attrs:{id:"what-about-domains-and-process-on-uncaughtexception"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#what-about-domains-and-process-on-uncaughtexception"}},[e._v("#")]),e._v(" What about domains and "),t("code",[e._v("process.on('uncaughtException')")]),e._v("?")]),e._v(" "),t("p",[e._v("Domains and the process-wide "),t("code",[e._v("'uncaughtException'")]),e._v(" event are primarily useful to attempt to handle or recover from unanticipated programmer errors. For all the reasons described above, this is strongly discouraged.")]),e._v(" "),t("h3",{attrs:{id:"specific-recommendations-for-writing-new-functions"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#specific-recommendations-for-writing-new-functions"}},[e._v("#")]),e._v(" Specific recommendations for writing new functions")]),e._v(" "),t("ul",[t("li",[e._v("Be clear about what your function does.\n"),t("ul",[t("li",[e._v("what arguments it expects")]),e._v(" "),t("li",[e._v("the types of each of those arguments")]),e._v(" "),t("li",[e._v("any additional constraints on those arguments (e.g., must be a valid IP address)")]),e._v(" "),t("li",[e._v("what operational errors callers should expect (including their "),t("code",[e._v("name")]),e._v("s)")]),e._v(" "),t("li",[e._v("how to handle operational errors (e.g., will they be thrown, passed to the callback, emitted on an event emitter, etc.)")]),e._v(" "),t("li",[e._v("the return value")])])]),e._v(" "),t("li",[e._v("Use "),t("code",[e._v("Error")]),e._v(" objects (or subclasses : "),t("code",[e._v("RangeError")]),e._v(", "),t("code",[e._v("ReferenceError")]),e._v(", "),t("code",[e._v("SyntaxError")]),e._v(", "),t("code",[e._v("TypeError")]),e._v(") for all errors, and implement the "),t("code",[e._v("Error")]),e._v(" contract ("),t("code",[e._v("name")]),e._v(", "),t("code",[e._v("message")]),e._v(" and "),t("code",[e._v("stack")]),e._v(" props at least).")]),e._v(" "),t("li",[e._v("Use the Error's "),t("code",[e._v("name")]),e._v(" property to distinguish errors programmatically (you don't need to subclass "),t("code",[e._v("Error")]),e._v(" for each error type).")]),e._v(" "),t("li",[e._v("Augment the "),t("code",[e._v("Error")]),e._v(" object with properties that explain details")]),e._v(" "),t("li",[e._v("If you pass a lower-level error to your caller, consider wrapping it instead. (wee "),t("a",{attrs:{href:"https://github.com/joyent/node-verror",target:"_blank",rel:"noopener noreferrer"}},[e._v("joyent/node-verror"),t("OutboundLink")],1),e._v(" on GitHub)")])])])}),[],!1,null,null,null);r.default=a.exports}}]);